sdk/src/main/java/com/google/cloud/dataflow/sdk/io/FileBasedSink.java

```diff
@@ -370,20 +372,42 @@ public void finalize(Iterable<FileResult> writerResults, PipelineOptions options
     }
 
     /**
+     * Use {@link #removeTemporaryFiles(Collection, PipelineOptions)} instead.
+     */
+    @Deprecated
+    protected final void removeTemporaryFiles(PipelineOptions options) throws IOException {
+      removeTemporaryFiles(Collections.<String>emptyList(), options);
+    }
+
+    /**
      * Removes temporary output files. Uses the temporary filename to find files to remove.
      *
+     * <p>Additionally, to partially mitigate the effects of filesystems with eventually-consistent
+     * directory matching APIs, takes a list of files that are known to exist - i.e. removes the
+     * union of the known files and files that the filesystem says exist in the directory.
+     *
+     * <p>Assumes that, if globbing had been strongly consistent, it would have matched all
+     * of knownFiles - i.e. on a strongly consistent filesystem, knownFiles can be ignored.
+     *
      * <p>Can be called from subclasses that override {@link FileBasedWriteOperation#finalize}.
      * <b>Note:</b>If finalize is overridden and does <b>not</b> rename or otherwise finalize
      * temporary files, this method will remove them.
      */
-    protected final void removeTemporaryFiles(PipelineOptions options) throws IOException {
+    protected final void removeTemporaryFiles(
+        Collection<String> knownFiles, PipelineOptions options) throws IOException {
       String pattern = buildTemporaryFilename(baseTemporaryFilename, "*");
       LOG.debug("Finding temporary bundle output files matching {}.", pattern);
       FileOperations fileOperations = FileOperationsFactory.getFileOperations(pattern, options);
       IOChannelFactory factory = IOChannelUtils.getFactory(pattern);
       Collection<String> matches = factory.match(pattern);
-      LOG.debug("{} temporary files matched {}", matches.size(), pattern);
-      LOG.debug("Removing {} files.", matches.size());
+      Set<String> allMatches = new HashSet<>(matches);
+      allMatches.addAll(knownFiles);
+      LOG.debug(
+          "Removing {} temporary files matching {} ({} matched glob, {} additional known files)",
```

Comment:
Not sure what you mean: the last parameter of the log message, Z, is defined as X - Y.